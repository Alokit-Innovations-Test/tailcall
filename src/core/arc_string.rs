use std::sync::Arc;

use headers::HeaderValue;
use schemars::gen::SchemaGenerator;
use schemars::schema::Schema;
use serde::{Deserialize, Deserializer, Serialize, Serializer};

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct ArcString(Arc<str>);

impl<'de> Deserialize<'de> for ArcString {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        String::deserialize(deserializer).map(ArcString::from)
    }
}

impl Serialize for ArcString {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.as_str().serialize(serializer)
    }
}

impl Default for ArcString {
    fn default() -> Self {
        ArcString("".into())
    }
}

impl AsRef<str> for ArcString {
    fn as_ref(&self) -> &str {
        &self.0
    }
}

impl From<&str> for ArcString {
    fn from(s: &str) -> Self {
        ArcString(s.into())
    }
}

impl From<&String> for ArcString {
    fn from(s: &String) -> Self {
        ArcString(s.clone().into())
    }
}

impl From<String> for ArcString {
    fn from(s: String) -> Self {
        ArcString(s.into())
    }
}

impl From<ArcString> for Vec<u8> {
    fn from(s: ArcString) -> Self {
        s.as_str().as_bytes().to_vec()
    }
}

impl TryFrom<ArcString> for HeaderValue {
    type Error = <HeaderValue as TryFrom<String>>::Error;

    fn try_from(s: ArcString) -> Result<Self, Self::Error> {
        s.as_str().parse()
    }
}

impl PartialEq<&str> for ArcString {
    fn eq(&self, other: &&str) -> bool {
        self.as_ref() == *other
    }
}

impl schemars::JsonSchema for ArcString {
    fn schema_name() -> String {
        <String as schemars::JsonSchema>::schema_name()
    }

    fn json_schema(gen: &mut SchemaGenerator) -> Schema {
        <String as schemars::JsonSchema>::json_schema(gen)
    }
}

impl std::fmt::Display for ArcString {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        self.as_str().fmt(f)
    }
}

impl ArcString {
    pub fn as_str(&self) -> &str {
        self.as_ref()
    }
}
