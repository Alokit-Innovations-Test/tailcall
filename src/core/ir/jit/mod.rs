mod builder;
mod context;
mod execute;
mod field_index;
mod model;
mod store;
mod synth;
