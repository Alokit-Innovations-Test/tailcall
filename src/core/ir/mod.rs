mod cache;
mod discriminator;
mod error;
mod eval;
mod evaluation_context;
mod graphql_operation_context;
mod io;
mod modify;
mod resolver_context_like;

use core::future::Future;
use std::fmt::Debug;
use std::pin::Pin;

use async_graphql_value::ConstValue;
pub use cache::*;
pub use discriminator::{Discriminator, TypeName};
pub use error::*;
pub use eval::*;
pub use evaluation_context::EvaluationContext;
pub use graphql_operation_context::GraphQLOperationContext;
pub use io::*;
pub use resolver_context_like::{EmptyResolverContext, ResolverContext, ResolverContextLike};
use strum_macros::Display;

use crate::core::blueprint::DynamicValue;
use crate::core::json::JsonLike;
use crate::core::serde_value_ext::ValueExt;

#[derive(Clone, Debug, Display)]
pub enum IR {
    Context(Context),
    Dynamic(DynamicValue),
    #[strum(to_string = "{0}")]
    IO(IO),
    Cache(Cache),
    Path(Box<IR>, Vec<String>),
    Protect(Box<IR>),
    Discriminate(Discriminator, Box<IR>),
}

#[derive(Clone, Debug)]
pub enum Context {
    Value,
    Path(Vec<String>),
    PushArgs { expr: Box<IR>, and_then: Box<IR> },
    PushValue { expr: Box<IR>, and_then: Box<IR> },
}

impl IR {
    pub fn and_then(self, next: Self) -> Self {
        IR::Context(Context::PushArgs { expr: Box::new(self), and_then: Box::new(next) })
    }

    pub fn with_args(self, args: IR) -> Self {
        IR::Context(Context::PushArgs { expr: Box::new(args), and_then: Box::new(self) })
    }
}

impl Eval for IR {
    #[tracing::instrument(skip_all, fields(otel.name = %self), err)]
    fn eval<'slf, 'ctx, Ctx: ResolverContextLike + Sync + Send>(
        &'slf self,
        ctx: &'ctx mut EvaluationContext<'slf, Ctx>,
    ) -> Pin<Box<dyn Future<Output = Result<ConstValue, EvaluationError>> + 'ctx + Send>> {
        Box::pin(async move {
            match self {
                IR::Context(op) => match op {
                    Context::Value => {
                        Ok(ctx.value().cloned().unwrap_or(async_graphql::Value::Null))
                    }
                    Context::Path(path) => Ok(ctx
                        .path_value(path)
                        .map(|a| a.into_owned())
                        .unwrap_or(async_graphql::Value::Null)),
                    Context::PushArgs { expr, and_then } => {
                        let args = expr.eval(ctx).await?;
                        let ctx = &mut ctx.with_args(args).clone();
                        and_then.eval(ctx).await
                    }
                    Context::PushValue { expr, and_then } => {
                        let value = expr.eval(ctx).await?;
                        let ctx = &mut ctx.with_value(value);
                        and_then.eval(ctx).await
                    }
                },
                IR::Path(input, path) => {
                    let inp = &input.eval(ctx).await?;
                    Ok(inp
                        .get_path(path)
                        .unwrap_or(&async_graphql::Value::Null)
                        .clone())
                }
                IR::Dynamic(value) => Ok(value.render_value(ctx)),
                IR::Protect(expr) => {
                    ctx.request_ctx
                        .auth_ctx
                        .validate(ctx.request_ctx)
                        .await
                        .to_result()?;
                    expr.eval(ctx).await
                }
                IR::IO(operation) => operation.eval(ctx).await,
                IR::Cache(cached) => cached.eval(ctx).await,
                IR::Discriminate(discriminator, expr) => expr.eval(ctx).await.and_then(|value| {
                    let type_name = discriminator.resolve_type(&value)?;

                    ctx.set_type_name(type_name);

                    Ok(value)
                }),
            }
        })
    }
}
