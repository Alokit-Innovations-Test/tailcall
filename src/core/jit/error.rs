use async_graphql::parser::types::OperationType;
use async_graphql::{ErrorExtensions, PathSegment, Pos, ServerError};
use serde::Serialize;
use thiserror::Error;

#[derive(Error, Debug, Clone, PartialEq, Eq)]
#[error("Error while building the plan")]
pub enum BuildError {
    #[error("Root Operation type not defined for {operation}")]
    RootOperationTypeNotDefined { operation: OperationType },
    #[error("ResolveInputError: {0}")]
    ResolveInputError(#[from] ResolveInputError),
    #[error(r#"Unknown operation named "{0}""#)]
    OperationNotFound(String),
    #[error("Operation name required in request")]
    OperationNameRequired,
}

#[derive(Error, Debug, Clone, PartialEq, Eq)]
#[error("Cannot resolve the input value")]
pub enum ResolveInputError {
    #[error("Variable `{0}` is not defined")]
    VariableIsNotFound(String),
}

#[derive(Error, Debug, Clone)]
pub enum ValidationError {
    // TODO: replace with sane error message. Right now, it's defined as is only for compatibility
    // with async_graphql error message for this case
    #[error(r#"internal: invalid value for scalar "{type_of}", expected "FieldValue::Value""#)]
    ScalarInvalid { type_of: String },
    #[error(r#"internal: invalid item for enum "{type_of}""#)]
    EnumInvalid { type_of: String },
    #[error("internal: non-null types require a return value")]
    ValueRequired,
}

#[derive(Debug, Clone, Error)]
pub enum Error {
    #[error("Build error: {0}")]
    BuildError(#[from] BuildError),
    #[error("ParseError: {0}")]
    ParseError(#[from] async_graphql::parser::Error),
    #[error(transparent)]
    IR(#[from] crate::core::ir::Error),
    #[error(transparent)]
    Validation(#[from] ValidationError),
}

impl ErrorExtensions for Error {
    fn extend(&self) -> async_graphql::Error {
        match self {
            Error::BuildError(error) => error.extend(),
            Error::ParseError(error) => error.extend(),
            Error::IR(error) => error.extend(),
            Error::Validation(error) => error.extend(),
        }
    }
}

pub type Result<A> = std::result::Result<A, Error>;

impl From<Error> for ServerError {
    fn from(val: Error) -> Self {
        // async_graphql::parser::Error has special conversion to ServerError
        if let Error::ParseError(error) = val {
            return error.into();
        }

        let extensions = val.extend().extensions;
        let mut server_error = ServerError::new(val.to_string(), None);

        server_error.extensions = extensions;

        server_error
    }
}

// impl From<Positioned<Error>> for Lift<ServerError> {
//     fn from(a: Positioned<Error>) -> Self {
//         (match a.node {
//             // async_graphql::parser::Error already has builtin positioning
//             Error::ParseError(error) => error.into(),
//             error => error.into_server_error_with_pos(Some(a.pos)),
//         })
//         .into()
//     }
// }

#[derive(Debug, Serialize)]
pub struct LocationError<Err> {
    pub error: Err,
    pub pos: Pos,
    pub path: Vec<PathSegment>,
}

impl From<LocationError<Error>> for ServerError {
    fn from(val: LocationError<Error>) -> Self {
        let extensions = val.error.extend().extensions;
        let mut server_error = ServerError::new(val.error.to_string(), Some(val.pos));

        server_error.extensions = extensions;

        // TODO: in order to be compatible with async_graphql path is only set for
        // validation errors here but in general we might consider setting it
        // for every error
        if let Error::Validation(_) = val.error {
            server_error.path = val.path;
        }

        server_error
    }
}
