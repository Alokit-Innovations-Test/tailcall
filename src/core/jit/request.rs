use std::collections::HashMap;

use async_graphql::parser::types::ExecutableDocument;
use derive_setters::Setters;
use serde::Deserialize;

use super::{Builder, Error, ExecutionPlan, Result};
use crate::core::blueprint::Blueprint;

#[derive(Debug, Deserialize, Setters)]
pub struct Request<Value> {
    #[serde(default)]
    pub query: String,
    #[serde(default, rename = "operationName")]
    pub operation_name: Option<String>,
    #[serde(default)]
    pub variables: HashMap<String, Value>,
    #[serde(default)]
    pub extensions: HashMap<String, Value>,
    pub document: ExecutableDocument,
}

impl TryFrom<async_graphql::Request> for Request<async_graphql_value::ConstValue> {
    type Error = Error;
    fn try_from(mut value: async_graphql::Request) -> Result<Self> {
        let executable_doc = value.parsed_query().map_err(Error::ServerError)?.to_owned();

        Ok(Self {
            query: value.query,
            operation_name: value.operation_name,
            variables: match value.variables.into_value() {
                async_graphql_value::ConstValue::Object(val) => {
                    HashMap::from_iter(val.into_iter().map(|(k, v)| (k.to_string(), v)))
                }
                _ => HashMap::new(),
            },
            extensions: value.extensions,
            document: executable_doc,
        })
    }
}

impl<Value> Request<Value> {
    pub fn try_new(&self, blueprint: &Blueprint) -> Result<ExecutionPlan> {
        let builder = Builder::new(blueprint, self.document.clone());
        builder.build().map_err(Error::BuildError)
    }
}

impl<A> TryFrom<&str> for Request<A> {
    type Error = Error;
    fn try_from(query: &str) -> std::result::Result<Self, Self::Error> {
        let document = async_graphql::parser::parse_query(query)?;
        Ok(Self {
            query: query.to_string(),
            operation_name: None,
            variables: HashMap::new(),
            extensions: HashMap::new(),
            document,
        })
    }
}
