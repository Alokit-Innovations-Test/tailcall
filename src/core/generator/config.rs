use std::path::Path;

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::core::config::{self, reader::ConfigReader};

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub enum InputSource {
    /// src: maintains the same src written in config.
    /// resolved_src: holds the correctly resolved src with respect to config.
    Config {
        src: String,
        resolved_src: Option<String>,
    },
    Import {
        src: String,
        resolved_src: Option<String>,
    },
}

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Input {
    #[serde(flatten)]
    pub source: InputSource,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Output {
    /// Controls the output format (graphql, json, yaml)
    pub format: config::Source,
    /// Specifies the output file name
    pub file: String,
}

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Schema {
    #[serde(default = "defaults::schema::query")]
    pub query: String,
    #[serde(default = "defaults::schema::mutation")]
    pub mutation: String,
}

impl Default for Schema {
    fn default() -> Self {
        Self {
            query: defaults::schema::query(),
            mutation: defaults::schema::mutation(),
        }
    }
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct GenerateOptions {
    #[serde(default)]
    pub schema: Schema,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Transform {
    // TODO: change types
    pub ambiguous_name_resolver: Option<serde_json::Value>,
    pub merge_types: Option<serde_json::Value>,
    pub js: Option<serde_json::Value>,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct GeneratorConfig {
    pub input: Vec<Input>,
    #[serde(default)]
    pub output: Output,
    #[serde(default)]
    pub generate: GenerateOptions,
    #[serde(default)]
    pub transform: Transform,
}

impl GeneratorConfig {
    /// Resolves all the paths present inside the GeneratorConfig.
    pub fn resolve_paths(mut self, file_path: &str) -> Self {
        let config_path = Path::new(file_path).parent().unwrap_or(Path::new(""));

        for input in self.input.iter_mut() {
            match &mut input.source {
                InputSource::Config { src, resolved_src }
                | InputSource::Import { src, resolved_src } => {
                    *resolved_src = Some(ConfigReader::resolve_path(src, Some(config_path)));
                }
            }
        }

        self.output.file = ConfigReader::resolve_path(&self.output.file, Some(config_path));

        self
    }
}

mod defaults {
    pub mod schema {
        pub fn query() -> String {
            "Query".to_string()
        }

        pub fn mutation() -> String {
            "Mutation".to_string()
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::GeneratorConfig;

    #[test]
    fn test_from_json() {
        let content = std::fs::read_to_string(tailcall_fixtures::generator::SIMPLE_JSON).unwrap();
        let config: GeneratorConfig = serde_json::from_str(&content).unwrap();

        assert_debug_snapshot!(&config);
    }

    #[test]
    fn test_resolve_paths() {
        let file_path = "../../../tailcall-fixtures/fixtures/generator/simple-json.json";
        let content = std::fs::read_to_string(tailcall_fixtures::generator::SIMPLE_JSON).unwrap();
        let config: GeneratorConfig = serde_json::from_str(&content).unwrap();
        let config = config.resolve_paths(file_path);
        assert_debug_snapshot!(&config);
    }
}
