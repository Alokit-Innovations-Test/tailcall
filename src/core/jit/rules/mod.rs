use async_graphql_value::Value;

use super::OperationPlan;
use crate::core::valid::{Valid, Validator};

mod query_complexity;
mod query_depth;

pub use query_complexity::QueryComplexity;
pub use query_depth::QueryDepth;

pub trait Rule {
    fn validate(&self, plan: &OperationPlan<Value>) -> Valid<(), String>;
}

pub trait RuleOps: Sized + Rule {
    fn pipe<Other: Rule>(self, other: Other) -> Pipe<Self, Other> {
        Pipe(self, other)
    }
}

impl<T: Rule> RuleOps for T {}

pub struct Pipe<A, B>(A, B);

impl<A, B> Rule for Pipe<A, B>
where
    A: Rule,
    B: Rule,
{
    fn validate(&self, plan: &OperationPlan<Value>) -> Valid<(), String> {
        self.0.validate(plan).and_then(|_| self.1.validate(plan))
    }
}
