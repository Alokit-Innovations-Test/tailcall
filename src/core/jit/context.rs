use async_graphql::{Name, ServerError};
use async_graphql_value::ConstValue;

use super::{Field, Nested, OperationPlan, Request};
use crate::core::ir::{ResolverContextLike, SelectionField};

/// Rust representation of the GraphQL context available in the DSL
#[derive(Debug, Clone)]
pub struct Context<'a, Input, Output> {
    request: &'a Request<Input>,
    value: Option<&'a Output>,
    args: Option<indexmap::IndexMap<Name, Input>>,
    // TODO: remove the args, since they're already present inside the fields and add support for
    // default values.
    field: &'a Field<Nested<Input>, Input>,
    plan: &'a OperationPlan<Input>,
}
impl<'a, Input: Clone, Output> Context<'a, Input, Output> {
    pub fn new(
        request: &'a Request<Input>,
        field: &'a Field<Nested<Input>, Input>,
        plan: &'a OperationPlan<Input>,
    ) -> Self {
        Self { request, value: None, args: None, field, plan }
    }

    pub fn with_value_and_field(
        &self,
        value: &'a Output,
        field: &'a Field<Nested<Input>, Input>,
    ) -> Self {
        Self {
            request: self.request,
            args: None,
            value: Some(value),
            field,
            plan: self.plan,
        }
    }

    pub fn with_field(&self, field: &'a Field<Nested<Input>, Input>) -> Self {
        let args = self.args.clone();

        Self {
            request: self.request,
            value: self.value,
            field,
            args,
            plan: self.plan,
        }
    }

    pub fn with_args(&self, args: indexmap::IndexMap<&str, Input>) -> Self {
        let mut map = indexmap::IndexMap::new();
        for (key, value) in args {
            map.insert(Name::new(key), value);
        }
        Self {
            request: self.request,
            value: self.value,
            args: Some(map),
            field: self.field,
            plan: self.plan,
        }
    }

    pub fn value(&self) -> Option<&Output> {
        self.value
    }
}

impl<'a> ResolverContextLike for Context<'a, ConstValue, ConstValue> {
    fn value(&self) -> Option<&ConstValue> {
        self.value
    }

    // TODO: make generic over type of stored values
    fn args(&self) -> Option<&indexmap::IndexMap<Name, ConstValue>> {
        self.args.as_ref()
    }

    fn field(&self) -> Option<SelectionField> {
        Some(SelectionField::from(self.field))
    }

    fn is_query(&self) -> bool {
        self.plan.is_query()
    }

    fn add_error(&self, error: ServerError) {
        self.plan.add_error(error.into())
    }
}

#[cfg(test)]
mod test {
    use super::Context;
    use crate::core::blueprint::Blueprint;
    use crate::core::config::{Config, ConfigModule};
    use crate::core::ir::ResolverContextLike;
    use crate::core::jit::{OperationPlan, Request};
    use crate::core::valid::Validator;

    fn setup(
        query: &str,
    ) -> (
        OperationPlan<async_graphql::Value>,
        Request<async_graphql::Value>,
    ) {
        let sdl = std::fs::read_to_string(tailcall_fixtures::configs::JSONPLACEHOLDER).unwrap();
        let config = Config::from_sdl(&sdl).to_result().unwrap();
        let blueprint = Blueprint::try_from(&ConfigModule::from(config)).unwrap();
        let request = Request::new(query);
        let plan = request.clone().create_plan(&blueprint).unwrap();
        (plan, request)
    }

    #[test]
    fn test_field() {
        let (plan, req) = setup("query {posts {id title}}");
        let field = plan.as_nested();
        let ctx: Context<async_graphql::Value, async_graphql::Value> =
            Context::new(&req, &field[0], &plan);
        insta::assert_debug_snapshot!(ctx.field().unwrap());
    }
}
