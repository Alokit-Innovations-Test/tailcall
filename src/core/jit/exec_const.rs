use std::sync::Arc;

use async_graphql_value::ConstValue;
use serde::Deserialize;
use super::context::Context;
use super::exec::{Executor, IRExecutor};
use super::{Error, OperationPlan, Request, Response, Result};
use crate::core::app_context::AppContext;
use crate::core::http::RequestContext;
use crate::core::ir::model::IR;
use crate::core::ir::EvalContext;
use crate::core::jit::synth::Synth;
use crate::core::json::JsonLike;

/// A specialized executor that executes with async_graphql::Value
pub struct ConstValueExecutor {
    // maybe we can convert it to generic val
    plan: OperationPlan<ConstValue>,
}

impl ConstValueExecutor {
    pub fn new<'a, Value: JsonLike<'a> + Deserialize<'a> + Clone>(request: &Request<ConstValue>, app_ctx: Arc<AppContext<Value>>) -> Result<Self> {
        Ok(Self { plan: request.create_plan(&app_ctx.blueprint)? })
    }

    pub async fn execute<'a, Value: JsonLike<'a> + Deserialize<'a> + Clone>(
        self,
        req_ctx: &'a RequestContext<Value>,
        request: Request<ConstValue>,
    ) -> Response<ConstValue, Error> {
        let exec = ConstValueExec::new(req_ctx);
        let plan = self.plan;
        // TODO: drop the clones in plan
        let vars = request.variables.clone();
        let exe = Executor::new(plan.clone(), exec);
        let store = exe.store(request).await;
        let synth = Synth::new(plan, store, vars);
        exe.execute(synth).await
    }
}

struct ConstValueExec<'a, Value> {
    req_context: &'a RequestContext<Value>,
}

impl<'a, Value: JsonLike<'a> + Deserialize<'a> + Clone> ConstValueExec<'a, Value> {
    pub fn new(ctx: &'a RequestContext<Value>) -> Self {
        Self { req_context: ctx }
    }
}

#[async_trait::async_trait]
impl<'ctx> IRExecutor for ConstValueExec<'ctx, async_graphql::Value> {
    type Input = ConstValue;
    type Output = ConstValue;
    type Error = Error;

    async fn execute<'a>(
        &'a self,
        ir: &'a IR<Self::Output>,
        ctx: &'a Context<'a, Self::Input, Self::Output>,
    ) -> Result<Self::Output> {
        let req_context = &self.req_context;
        let mut ctx = EvalContext::new(req_context, ctx);
        Ok(ir.eval(&mut ctx).await?)
    }
}
