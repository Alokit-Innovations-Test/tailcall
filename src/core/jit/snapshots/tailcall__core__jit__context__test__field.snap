---
source: src/core/jit/context.rs
expression: ctx.field().unwrap()
---
SelectionField {
    name: "posts",
    args: [],
    directive: Ok(
        [],
    ),
    selection_set: [
        SelectionField {
            name: "id",
            args: [],
            directive: Ok(
                [],
            ),
            selection_set: [],
        },
        SelectionField {
            name: "title",
            args: [],
            directive: Ok(
                [],
            ),
            selection_set: [],
        },
    ],
}
